ras_party/controllers/account_controller.py

```diff
@@ -0,0 +1,461 @@
+
+from ras_party.clients.oauth_client import OauthClient
+from ras_party.controllers.error_decorator import translate_exceptions
+from ras_party.controllers.gov_uk_notify import GovUkNotify
+from ras_party.controllers.queries import query_business_by_party_uuid, query_respondent_by_email, \
+    query_respondent_by_party_uuid
+from ras_party.controllers.validate import Validator, IsUuid, Exists
+from ras_party.models.models import Respondent, RespondentStatus, BusinessRespondent, PendingEnrolment, \
+    Enrolment, EnrolmentStatus
+from ras_party.support.public_website import PublicWebsite
+from ras_party.support.requests_wrapper import Requests
+from ras_party.support.session_decorator import with_db_session
+from ras_party.support.transactional import transactional
+from ras_party.support.util import build_url
+from ras_party.support.verification import decode_email_token
+
+log = get_logger()
+
+NO_RESPONDENT_FOR_PARTY_ID = 'There is no respondent with that party ID '
+EMAIL_VERIFICATION_SENT = 'A new verification email has been sent'
+
+
+@translate_exceptions
+@transactional
+@with_db_session
+def post_respondent(party, tran, session):
+    """
+    This function is quite complicated, as it performs a number of steps to complete a new respondent enrolment:
+    1. Validate the enrolment code and email address
+    2. Register the respondent as a new user with the remote OAuth2 service.
+    3. Lookup the case context from the case service
+    4. Lookup the collection exercise from the collection exercise service
+    5. Lookup the survey from the survey service
+    6. Generate an email verification token, and an email verification url
+    7. Invoke the email verification
+    8. Update the database with the new respondent, establishing an association to the business, and an entry
+       in the enrolment table (along with the survey id / survey name)
+    9. Post a case event to the case service to notify that a new respondent has been created
+    10. If something goes wrong after step 1, attempt to perform a compensating action to remove the OAuth2 user
+       (this is currently mocked out as the OAuth2 server doesn't implement an endpoint to achieve this)
+    """
+
+    # Validation, curation and checks
+    expected = ('emailAddress', 'firstName', 'lastName', 'password', 'telephone', 'enrolmentCode')
+
+    v = Validator(Exists(*expected))
+    if 'id' in party:
+        # Note: there's not strictly a requirement to be able to pass in a UUID, this is currently supported to
+        # aid with testing.
+        log.debug("'id' in respondent post message. Adding validation rule IsUuid")
+        v.add_rule(IsUuid('id'))
+
+    if not v.validate(party):
+        raise RasError(v.errors, 400)
+
+    iac = request_iac(party['enrolmentCode'])
+    if not iac.get('active'):
+        raise RasError("Enrolment code is not active.", status_code=400)
+
+    existing = query_respondent_by_email(party['emailAddress'], session)
+    if existing:
```

Comment:
could just do `if query_respondent_by_email(party['emailAddress'], session)` since you only use existing once